### PR TITLE
pyenv: init at 2.3.21

### LINKED ISSUE #238898: Package request: pyenv
**Project description**

pyenv lets you easily switch between multiple versions of Python. It's simple, unobtrusive, and follows the UNIX tradition of single-purpose tools that do one thing well.

It is a critical application for Python development.

**Metadata**

* homepage URL: https://github.com/pyenv/pyenv
* source URL: https://github.com/pyenv/pyenv
* license: MIT
* platforms: ALL

### PR DESCRIPTION
**Pull request by tjni**
###### Description of changes

[pyenv](https://github.com/pyenv/pyenv) is a collection of bash scripts that help with managing Python versions.

I have only tested some basic functionality on aarch64-darwin. Will resolve https://github.com/NixOS/nixpkgs/issues/238898.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/pyenv/default.nix [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  configureScript = "src/configure";
+
+  makeFlags = ["-C" "src"];
+
+  installPhase = ''
+    mkdir -p "$out"
+    cp -R bin "$out/bin"
+    cp -R libexec "$out/libexec"
+    cp -R plugins "$out/plugins"
+  '';
```
Comment (natsukium) [inline pkgs/development/tools/pyenv/default.nix:26]:
```suggestion
  installPhase = ''
    runHook preInstall
    
    mkdir -p "$out"
    cp -R bin "$out/bin"
    cp -R libexec "$out/libexec"
    cp -R plugins "$out/plugins"
    
    runHook postInstall
  '';
```
please add runHook

Comment (tjni) [reply pkgs/development/tools/pyenv/default.nix]: Done, thanks!

File: pkgs/development/tools/pyenv/default.nix [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  version = "2.3.21";
+
+  src = fetchFromGitHub {
+    owner = "pyenv";
+    repo = "pyenv";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-+rMkbrRSQBHX9qdifIpdLmQ8Uk39ott0xyJTD/mb+9w=";
+  };
+
+  configureScript = "src/configure";
```
Comment (natsukium) [inline pkgs/development/tools/pyenv/default.nix:17]: The configure script [has shebangs](https://github.com/pyenv/pyenv/blob/a2dff4809b8a7c387f26f6f5760c97959fdc940c/src/configure#L1) and cannot be built on Linux as is, so it is necessary to run patchShebangs.

File: pkgs/development/tools/pyenv/default.nix [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+  makeFlags = ["-C" "src"];
+
+  installPhase = ''
```
Comment (natsukium) [inline pkgs/development/tools/pyenv/default.nix:21]: We can use `installShellFiles` to install completions and man pages that the upstream provides.

File: pkgs/development/tools/pyenv/default.nix [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    cp -R libexec "$out/libexec"
+    cp -R plugins "$out/plugins"
+
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    description = "Simple Python version management";
+    homepage = "https://github.com/pyenv/pyenv";
+    changelog = "https://github.com/pyenv/pyenv/blob/master/CHANGELOG.md";
```
Comment (natsukium) [inline pkgs/development/tools/pyenv/default.nix:35]:
```suggestion
    changelog = "https://github.com/pyenv/pyenv/blob/${src.rev}/CHANGELOG.md";
```

**Comment by natsukium** [comment]
It seems that we need to patch around the python build to actually use it on NixOS.
However, those using NixOS probably don't need pyenv, so it's fine now.

File: pkgs/development/tools/pyenv/default.nix (hunk 2 above)
Comment (tjni) [reply pkgs/development/tools/pyenv/default.nix:26]: Thank you for the review and the help. This should be fixed now.

**Review by natsukium — APPROVED** [review]

**Comment by StepBroBD** [comment]
Related: https://github.com/NixOS/nixpkgs/pull/94769

**Comment by tjni** [comment]
Thank you for pointing me to that attempt, I had not seen it before. Its implementation doesn't look too different from mine except that the (major) version is different.

It prompted me to do a more thorough end-to-end test, and I'm able to get it to work on my macOS system. The default `PYENV_ROOT` seems to be `~/.pyenv`, and indeed Python is installed there and can be used.

**Pull request merged**